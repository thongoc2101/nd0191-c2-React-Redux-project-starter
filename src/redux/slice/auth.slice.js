import { createSlice } from "@reduxjs/toolkit";
import { AuthActions } from "../actions/auth.action";

const initialState = {
    userInfo: null,
    users: null,
    isLogin: false
};

export const AuthSlice = createSlice({
    name: 'auth',
    initialState: initialState,
    reducers: {
        login: (state, action) => {
            const user = Object.values(state.users).find(user => user.id === action.payload.username && user.password === action.payload.password);
            if (user) {
                state.isLogin = true;
                state.userInfo = user;
            } else {
                state.isLogin = false;
                state.userInfo = null;
            }
        },
        logout: (state) => {
            state.isLogin = false;
            state.userInfo = null;
        },

        setQuestionUser: (state, action) => {
            const userItem = Object.values(state.users).find(user => user.id === action.payload.author);
            userItem.questions.push(action.payload.id);

            state.userInfo = userItem;
            state.users = {
                ...state.users,
                [userItem.id]: userItem
            }
        },

        setUserAnswer: (state, action) => {
            const authedUser = action.payload.authedUser;
            const answer = action.payload.answer;
            const qid = action.payload.id;

            state.users = {
                ...state.users,
                [authedUser]: {
                  ...state.users[authedUser],
                  answers: {
                    ...state.users[authedUser].answers,
                    [qid]: answer
                  }
                }
              }
        }
    },
    extraReducers: (builder) => {
        // Add extra reducers here
        builder.addCase(AuthActions.getUsers.fulfilled, (state, action) => {
            state.users = action.payload;
        });
    }
});

export const { login, logout, setQuestionUser, setUserAnswer } = AuthSlice.actions;