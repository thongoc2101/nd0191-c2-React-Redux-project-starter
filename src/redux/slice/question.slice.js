import { createSlice } from "@reduxjs/toolkit";
import { QuestionActions } from "../actions/question.action";

const initialState = {
    questions: null,
};

export const QuestionSlice = createSlice({
    name: 'questions',
    initialState: initialState,
    reducers: {
        setAnsweredUser: (state, action) => {
            const qid = action.payload.id;
            const answer = action.payload.answer;
            const authedUser = action.payload.authedUser;

            state.questions = {
                ...state.questions,
                [qid]: {
                  ...state.questions[qid],
                  [answer]: {
                    ...state.questions[qid][answer],
                    votes: state.questions[qid][answer].votes.push(authedUser)
                  }
                }
              }
        }
    },
    extraReducers: (builder) => {
        // Add extra reducers here
        builder.addCase(QuestionActions.getQuestions.fulfilled, (state, action) => {
            state.questions = action.payload;
        });
    }
});

export const { setAnsweredUser } = QuestionSlice.actions;