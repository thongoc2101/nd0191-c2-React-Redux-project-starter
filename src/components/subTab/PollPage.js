import { useDispatch, useSelector } from "react-redux";
import { useNavigate, useParams } from "react-router-dom";
import { QuestionActions } from "../../redux/actions/question.action";
import { useCallback, useEffect, useState } from "react";
import { setAnsweredUser } from "../../redux/slice/question.slice";
import { setUserAnswer } from "../../redux/slice/auth.slice";

const PollPage = () => {
    const dispatch = useDispatch();
    const navigate = useNavigate();

    const params = useParams();
    const questionId = params.id;

    const { userInfo } = useSelector((state) => state.auth);

    const [question, setQuestion] = useState(null);
    const [hasVotedForOptionOne, setHasVotedForOptionOne] = useState(false);
    const [hasVotedForOptionTwo, setHasVotedForOptionTwo] = useState(false);
    const [hasVoted, setHasVoted] = useState(false);

    const getQuestionsInfo = useCallback(() => {
        dispatch(QuestionActions.getQuestions()).then((res) => {
            const question = Object.values(res.payload).find((question) => question.id === questionId);
            if (!question) navigate({ pathname: '/404' })

            setQuestion(question);

            const hasVotedForOptionOneTemp = question?.optionOne?.votes.includes(userInfo.id);
            const hasVotedForOptionTwoTemp = question?.optionTwo?.votes.includes(userInfo.id);

            setHasVotedForOptionOne(hasVotedForOptionOneTemp);
            setHasVotedForOptionTwo(hasVotedForOptionTwoTemp);
            setHasVoted(hasVotedForOptionOneTemp || hasVotedForOptionTwoTemp);
        });
    }, [dispatch, questionId, userInfo.id, navigate]);

    const handleChangeOption = (e, optionChoose) => {
        e.preventDefault();
        dispatch(QuestionActions.saveQuestionAnswer({
            qid: question.id, answer: optionChoose, authedUser: userInfo.id
        })).then((res) => {
            if (res) {
                dispatch(setAnsweredUser({ id: question.id, answer: optionChoose, authedUser: userInfo.id }));
                dispatch(setUserAnswer({ id: question.id, answer: optionChoose, authedUser: userInfo.id }));
                getQuestionsInfo();
            }
        })
    };

    const calcVotes = (option, question) => {
        const numberVotesTotal = question.optionOne.votes.length + question.optionTwo.votes.length;
        switch (option) {
            case "optionOne":
                return question.optionOne.votes.length / numberVotesTotal * 100 + "%";
            case "optionTwo":
                return question.optionTwo.votes.length / numberVotesTotal * 100 + "%";
            default:
                return "";
        }
    };

    const renderTemplateVoted = (question, numberVotes, option) => {
        return (
            <>
                {!hasVoted && <p className="bg-sky-500 rounded-md">Click</p>}
                {hasVoted && <p className="text-xs">Votes: {numberVotes} ({calcVotes(option, question)})</p>}
            </>
        )
    }

    // init data
    useEffect(() => {
        getQuestionsInfo();
    }, [getQuestionsInfo]);

    return (
        <div className="poll-page-information">
            <h1 className="text-2xl font-bold mt-9">Poll by {question?.author}</h1>

            <div className="flex justify-center">
                <img src={userInfo.avatarURL} alt="" className="h-20 w-20" />
            </div>

            <div className="flex justify-center">
                <h2 className="text-2xl font-bold mt-6">Would you rather?</h2>
            </div>

            <div className="grid grid-cols-2 gap-4 mt-4">
                {/* button option one */}
                <button
                    onClick={(e) => handleChangeOption(e, 'optionOne')}
                    disabled={hasVoted}
                    className={"p-2 rounded-xl bg-white border border-solid" + (hasVotedForOptionOne ? " !bg-lime-500" : "")}>
                    <p className="mb-2">{question?.optionOne?.text}</p>
                    {renderTemplateVoted(question, question?.optionOne?.votes.length, "optionOne")}
                </button>

                {/* button option two */}
                <button
                    onClick={(e) => handleChangeOption(e, 'optionTwo')}
                    disabled={hasVoted}
                    className={"p-2 rounded-xl bg-white border border-solid" + (hasVotedForOptionTwo ? " !bg-lime-500" : "")}>
                    <p className="mb-2">{question?.optionTwo?.text}</p>
                    {renderTemplateVoted(question, question?.optionTwo?.votes.length, "optionTwo")}
                </button>
            </div>
        </div>
    );
}

export default PollPage;