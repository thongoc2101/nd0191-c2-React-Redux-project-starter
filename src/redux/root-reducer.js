import { combineReducers } from "@reduxjs/toolkit";
import { AuthSlice } from "./slice/auth.slice";
import { QuestionSlice } from "./slice/question.slice";

export const rootReducer = combineReducers({
    auth: AuthSlice.reducer,
    questions: QuestionSlice.reducer,
});