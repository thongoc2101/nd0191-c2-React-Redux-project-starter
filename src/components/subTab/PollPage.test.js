import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { store } from "../../redux/store";
import { AuthActions } from '../../redux/actions/auth.action';
import { login, setUserAnswer } from '../../redux/slice/auth.slice';
import PollPage from './PollPage';
import { QuestionActions } from '../../redux/actions/question.action';
import { setAnsweredUser } from '../../redux/slice/question.slice';
import { act } from 'react-dom/test-utils';

describe("PollPage", () => {
    it("should render the component when logged in", async () => {
        await store.dispatch(AuthActions.getUsers());
        await store.dispatch(login({ username: 'tylermcginnis', password: 'abc321' }))
        await store.dispatch(QuestionActions.getQuestions());

        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <PollPage />
                </BrowserRouter>
            </Provider>
        );

        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
    });

    it("should save question answer when click votes", async () => {
        await store.dispatch(AuthActions.getUsers());
        await store.dispatch(login({ username: 'tylermcginnis', password: 'abc321' }));
        await store.dispatch(QuestionActions.getQuestions());

        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <PollPage />
                </BrowserRouter>
            </Provider>
        );

        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();

        const pollPageHeadingElement = component.getByTestId("pollPage-header");
        const submitOptionOneButtonElement = component.getByTestId("submit-optionOne");
        expect(submitOptionOneButtonElement).toBeInTheDocument();

        expect(submitOptionOneButtonElement.textContent).toBe("Click");
        if (submitOptionOneButtonElement) {
            fireEvent.click(submitOptionOneButtonElement);
        }

        await act(async () => {
            store.dispatch(QuestionActions.saveQuestionAnswer({ qid: 'xj352vofupe1dqz9emx13r', answer: 'optionOne', authedUser: "sarahedo" }));
            store.dispatch(setAnsweredUser({ id: 'xj352vofupe1dqz9emx13r', answer: 'optionOne', authedUser: "sarahedo" }));
            store.dispatch(setUserAnswer({ id: 'xj352vofupe1dqz9emx13r', answer: 'optionOne', authedUser: "sarahedo" }));
            store.dispatch(QuestionActions.getQuestions());
        });

        expect(pollPageHeadingElement).toBeInTheDocument();
    });
});
