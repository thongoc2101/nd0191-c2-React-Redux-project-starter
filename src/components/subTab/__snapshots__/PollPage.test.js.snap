// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollPage should render the component when logged in 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="poll-page-information"
      >
        <h1
          class="text-2xl font-bold mt-9"
          data-testid="pollPage-header"
        >
          Poll by 
        </h1>
        <div
          class="flex justify-center"
        >
          <img
            alt=""
            class="h-20 w-20"
          />
        </div>
        <div
          class="flex justify-center"
        >
          <h2
            class="text-2xl font-bold mt-6"
          >
            Would you rather?
          </h2>
        </div>
        <div
          class="grid grid-cols-2 gap-4 mt-4"
        >
          <button
            class="p-2 rounded-xl bg-white border border-solid"
          >
            <p
              class="mb-2"
            />
            <div>
              <p
                class="bg-sky-500 rounded-md"
                data-testid="submit-optionOne"
              >
                Click
              </p>
            </div>
          </button>
          <button
            class="p-2 rounded-xl bg-white border border-solid"
          >
            <p
              class="mb-2"
            />
            <div>
              <p
                class="bg-sky-500 rounded-md"
                data-testid="submit-optionTwo"
              >
                Click
              </p>
            </div>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="poll-page-information"
    >
      <h1
        class="text-2xl font-bold mt-9"
        data-testid="pollPage-header"
      >
        Poll by 
      </h1>
      <div
        class="flex justify-center"
      >
        <img
          alt=""
          class="h-20 w-20"
        />
      </div>
      <div
        class="flex justify-center"
      >
        <h2
          class="text-2xl font-bold mt-6"
        >
          Would you rather?
        </h2>
      </div>
      <div
        class="grid grid-cols-2 gap-4 mt-4"
      >
        <button
          class="p-2 rounded-xl bg-white border border-solid"
        >
          <p
            class="mb-2"
          />
          <div>
            <p
              class="bg-sky-500 rounded-md"
              data-testid="submit-optionOne"
            >
              Click
            </p>
          </div>
        </button>
        <button
          class="p-2 rounded-xl bg-white border border-solid"
        >
          <p
            class="mb-2"
          />
          <div>
            <p
              class="bg-sky-500 rounded-md"
              data-testid="submit-optionTwo"
            >
              Click
            </p>
          </div>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`PollPage should save question answer when click votes 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="poll-page-information"
      >
        <h1
          class="text-2xl font-bold mt-9"
          data-testid="pollPage-header"
        >
          Poll by 
        </h1>
        <div
          class="flex justify-center"
        >
          <img
            alt=""
            class="h-20 w-20"
          />
        </div>
        <div
          class="flex justify-center"
        >
          <h2
            class="text-2xl font-bold mt-6"
          >
            Would you rather?
          </h2>
        </div>
        <div
          class="grid grid-cols-2 gap-4 mt-4"
        >
          <button
            class="p-2 rounded-xl bg-white border border-solid"
          >
            <p
              class="mb-2"
            />
            <div>
              <p
                class="bg-sky-500 rounded-md"
                data-testid="submit-optionOne"
              >
                Click
              </p>
            </div>
          </button>
          <button
            class="p-2 rounded-xl bg-white border border-solid"
          >
            <p
              class="mb-2"
            />
            <div>
              <p
                class="bg-sky-500 rounded-md"
                data-testid="submit-optionTwo"
              >
                Click
              </p>
            </div>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="poll-page-information"
    >
      <h1
        class="text-2xl font-bold mt-9"
        data-testid="pollPage-header"
      >
        Poll by 
      </h1>
      <div
        class="flex justify-center"
      >
        <img
          alt=""
          class="h-20 w-20"
        />
      </div>
      <div
        class="flex justify-center"
      >
        <h2
          class="text-2xl font-bold mt-6"
        >
          Would you rather?
        </h2>
      </div>
      <div
        class="grid grid-cols-2 gap-4 mt-4"
      >
        <button
          class="p-2 rounded-xl bg-white border border-solid"
        >
          <p
            class="mb-2"
          />
          <div>
            <p
              class="bg-sky-500 rounded-md"
              data-testid="submit-optionOne"
            >
              Click
            </p>
          </div>
        </button>
        <button
          class="p-2 rounded-xl bg-white border border-solid"
        >
          <p
            class="mb-2"
          />
          <div>
            <p
              class="bg-sky-500 rounded-md"
              data-testid="submit-optionTwo"
            >
              Click
            </p>
          </div>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
